### PR TITLE
optimize: align up resource reservation based on kubelet

### PR DESCRIPTION
**Pull request by justadogistaken**
refer: https://github.com/volcano-sh/volcano/issues/1885

Taking system-reserved into consideration.
But it may be not enough, this still fails to take "evictionReservation" into consideration. Because it relys on `ExperimentalNodeAllocatableIgnoreEvictionThreshold` in kubelet flag. And I haven't find a way to get this flag.

```
var hardEvictionThresholds []evictionapi.Threshold
		// If the user requested to ignore eviction thresholds, then do not set valid values for hardEvictionThresholds here.
		if !s.ExperimentalNodeAllocatableIgnoreEvictionThreshold {
			hardEvictionThresholds, err = eviction.ParseThresholdConfig([]string{}, s.EvictionHard, nil, nil, nil)
			if err != nil {
				return err
			}
		}
```

discusstion:
should we take "hardEvictionThresholds" into consideration by default? And leave the flag "ExperimentalNodeAllocatableIgnoreEvictionThreshold" alone (because we can't get this option value).
I think it is reasonable, because numa-aware scheduling should avoid pod rejected by kubelet.
Leaving "ExperimentalNodeAllocatableIgnoreEvictionThreshold" alone and Taking "hardEvictionThresholds" into calculation whenerer could help avoid pod rejection on host.

### CODE REVIEW

**Comment by justadogistaken** [comment]
ptal @huone1

**Comment by k82cn** [comment]
@justadogistaken , thanks very much for your PR, that's reasonable to me; would you help to seperate this PR into two commits, one commit includes diff in volcano/pkg, the other commit incloude the diff in vendors?

File: pkg/numatopo/update.go [hunk 1]
```diff
@@ -34,15 +34,14 @@ import (
...
-		isChange = true
-	}
-
-	if TopoInfoUpdate(opt) {
-		isChange = true
+	klConfig, err := GetKubeletConfigFromLocalFile(opt.KubeletConf)
+	if err != nil {
+		klog.Errorf("failed to get kubelet configuration, err: %v", err)
+	} else {
+		isChange = isChange || TryUpdatingResourceReservation(klConfig, opt.ResReserved)
```
Comment (huone1) [inline pkg/numatopo/update.go]: the  isChange value hasn't changed

**Comment by huone1** [comment]
> @justadogistaken , thanks very much for your PR, that's reasonable to me; would you help to seperate this PR into two commits, one commit includes diff in volcano/pkg, the other commit incloude the diff in vendors?

Is the vendor  needed to add ？？

**Comment by justadogistaken** [comment]
> @justadogistaken , thanks very much for your PR, that's reasonable to me; would you help to seperate this PR into two commits, one commit includes diff in volcano/pkg, the other commit incloude the diff in vendors?

Alright, I will split them. And maybe we need to ignore the vendors? @huone1

**Comment by huone1** [comment]
I review the code about ExperimentalNodeAllocatableIgnoreEvictionThreshold， I also didn't find the way to get it for resource-exporter
We can make the following constraints：
if evictionHard is setted in kubeconfig， ExperimentalNodeAllocatableIgnoreEvictionThreshold must be false

**Comment by justadogistaken** [comment]
> I review the code about ExperimentalNodeAllocatableIgnoreEvictionThreshold， I also didn't find the way to get it for resource-exporter We can make the following constraints： if evictionHard is setted in kubeconfig， ExperimentalNodeAllocatableIgnoreEvictionThreshold must be false

deal!

**Comment by justadogistaken** [comment]
hardEviction is nothing to do with cpu.

**Comment by huone1** [comment]
> hardEviction is nothing to do with cpu.

yep， you need to  consider it with the memory topolofy, and it is unnecessary for CPU topology

**Comment by huone1** [comment]
Is the  test  ok for the PR？

**Comment by justadogistaken** [comment]
> 

I tested it. And it's ok. I could supplement some unit tests later.
I haven't consider memery yet, because volcano does not have memory numa aware currently. But the pr will calculate memory reservation.
I think it's ok for now.

**Comment by justadogistaken** [comment]
@huone1  is "res-reserved" of resource-exporter same as "reserved-cpus" and "reserved-memory" of kubernetes? Why didn't we just take "reserved-cpus" from kubelet configutation. Instead, exporting "res-reserved" to user?

**Comment by justadogistaken** [comment]
How about that we fix the absence of “SystemReserved” in this pr(already tested). And then we complement the memory topology. I guess the work of this pr is fine.

**Comment by justadogistaken** [comment]
@huone1 Do you have time for reviewing this work?

**Comment by huone1** [comment]
Everything else is OK except the import file needs to be sorted out

**Comment by justadogistaken** [comment]
> Everything else is OK except the import file needs to be sorted out

Which file needs to be sorted? Could you help to point that out? I applied the file watcher for sorting imports, but failed to find unsorted imports

File: main.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ import (
 	"os"
 	"time"
 
+	"volcano.sh/resource-exporter/pkg/machineinfo"
```
Comment (Thor-wl) [inline main.go]: Pls keep the imported packages in order. In general, go package first, third package such as k8s second, packages in volcano last.

Comment (justadogistaken) [reply main.go]: Is there any sort rule file or tool? Or we need to adjust it manually?

Comment (Thor-wl) [reply main.go]: > Is there any sort rule file or tool? Or we need to adjust it manually?

It's by manually temporarily.

**Comment by justadogistaken** [comment]
Chcke the following part in k8s may help understand what's this work for.
Kubelet calculates the resourece reservation with "kube-reserved", "system-reserved", "eviction-reserved". But we just have  "kube-system" into calcaculation. In this work,  I put "system-reserved", "eviction-reserved" into calculation. But haven't considered the memory case (we don't need it right now). After this work merged, I will go on fill up memory topology info into numa crd.
```
// GetNodeAllocatableReservation returns amount of compute or storage resource that have to be reserved on this node from scheduling.
func (cm *containerManagerImpl) GetNodeAllocatableReservation() v1.ResourceList {
	evictionReservation := hardEvictionReservation(cm.HardEvictionThresholds, cm.capacity)
	result := make(v1.ResourceList)
	for k := range cm.capacity {
		value := resource.NewQuantity(0, resource.DecimalSI)
		if cm.NodeConfig.SystemReserved != nil {
			value.Add(cm.NodeConfig.SystemReserved[k])
		}
		if cm.NodeConfig.KubeReserved != nil {
			value.Add(cm.NodeConfig.KubeReserved[k])
		}
		if evictionReservation != nil {
			value.Add(evictionReservation[k])
		}
		if !value.IsZero() {
			result[k] = *value
		}
	}
	return result
}
```

**Review by Thor-wl — APPROVED** [review]

**Pull request merged**